File: dashboard/src/service/api/audit.ts
import { keepPreviousData, useQuery } from '@tanstack/react-query'
import { $fetch, fetcher } from '@/service/http'

export type AuditResult = 'success' | 'failure'

export interface AuditLog {
  id: number
  actor_id?: number | null
  actor_username?: string | null
  source_ip?: string | null
  action: string
  resource: string
  resource_id?: string | null
  before?: Record<string, unknown> | unknown[] | null
  after?: Record<string, unknown> | unknown[] | null
  result: AuditResult
  detail?: string | null
  created_at: string
}

export interface AuditLogQuery {
  search?: string
  actor?: string
  action?: string
  resource?: string
  result?: AuditResult
  start?: string
  end?: string
  offset?: number
  limit?: number
}

export interface AuditLogsResponse {
  logs: AuditLog[]
  total: number
  offset: number
  limit: number
}

const cleanParams = (params: AuditLogQuery) => Object.fromEntries(Object.entries(params).filter(([, value]) => value !== undefined && value !== ''))

export const getAuditLogs = (params: AuditLogQuery) => fetcher<AuditLogsResponse>('/api/audit', { params: cleanParams(params) })

export const getAuditLog = (id: number) => fetcher<AuditLog>(`/api/audit/${id}`)

export const useAuditLogs = (params: AuditLogQuery, enabled = true) =>
  useQuery({
    queryKey: ['audit-logs', params],
    queryFn: () => getAuditLogs(params),
    placeholderData: keepPreviousData,
    enabled,
    staleTime: 10_000,
  })

export const useAuditLog = (id: number | null, enabled = true) =>
  useQuery({
    queryKey: ['audit-log', id],
    queryFn: () => getAuditLog(id as number),
    enabled: enabled && id !== null,
  })

export const downloadAuditCsv = async (params: AuditLogQuery) => {
  const blob = await $fetch<Blob>('/api/audit/export', {
    params: cleanParams(params),
    responseType: 'blob',
  })
  const href = URL.createObjectURL(blob)
  const anchor = document.createElement('a')
  anchor.href = href
  anchor.download = `audit-logs-${new Date().toISOString().slice(0, 10)}.csv`
  document.body.appendChild(anchor)
  anchor.click()
  anchor.remove()
  URL.revokeObjectURL(href)
}

File: app/models/audit.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


class AuditLogResponse(BaseModel):
    id: int
    actor_id: int | None = None
    actor_username: str | None = None
    source_ip: str | None = None
    action: str
    resource: str
    resource_id: str | None = None
    before: dict[str, Any] | list[Any] | None = None
    after: dict[str, Any] | list[Any] | None = None
    result: str
    detail: str | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class AuditLogsResponse(BaseModel):
    logs: list[AuditLogResponse]
    total: int
    offset: int
    limit: int


class AuditLogQuery(BaseModel):
    search: str | None = Field(default=None, max_length=256)
    actor: str | None = Field(default=None, max_length=128)
    action: str | None = Field(default=None, max_length=64)
    resource: str | None = Field(default=None, max_length=64)
    result: Literal["success", "failure"] | None = None
    start: datetime | None = None
    end: datetime | None = None
    offset: int = Field(default=0, ge=0)
    limit: int = Field(default=50, ge=1, le=200)

    @field_validator("start", "end")
    @classmethod
    def require_timezone(cls, value: datetime | None) -> datetime | None:
        if value is not None and value.tzinfo is None:
            raise ValueError("date filters must include a timezone")
        return value

    @field_validator("end")
    @classmethod
    def validate_range(cls, value: datetime | None, info) -> datetime | None:
        start = info.data.get("start")
        if value is not None and start is not None and value < start:
            raise ValueError("end must be greater than or equal to start")
        return value
